elasticsearch/_async/helpers.py

```diff
@@ -439,24 +440,47 @@ async def async_reindex(
         :func:`~elasticsearch.helpers.async_scan`
     :arg bulk_kwargs: additional kwargs to be passed to
         :func:`~elasticsearch.helpers.async_bulk`
+    :arg op_type: Explicit operation type.
     """
     target_client = client if target_client is None else target_client
     docs = async_scan(
         client, query=query, index=source_index, scroll=scroll, **scan_kwargs
     )
 
-    async def _change_doc_index(hits, index):
+    async def _change_doc_index(hits, index, op_type):
         async for h in hits:
             h["_index"] = index
+            if op_type is not None:
+                h["_op_type"] = op_type
             if "fields" in h:
                 h.update(h.pop("fields"))
             yield h
 
     kwargs = {"stats_only": True}
     kwargs.update(bulk_kwargs)
+
+    is_data_stream = False
+    try:
+        # Verify if the target_index is data stream or index
+        data_streams = await target_client.indices.get_data_stream(target_index)
+    except NotFoundError:
+        # If its not data stream, might be index
+        pass
+    else:
+        for stream in data_streams["data_streams"]:
+            if target_index == stream["name"]:
+                is_data_stream = True
```

Comment:
`is_data_stream = any(data_stream["name"] == target_index for data_stream in data_streams["data_streams"])`